lib/erl/src/thrift_compact_protocol.erl

```diff
@@ -363,26 +365,32 @@ read(This0, double) ->
 % returns a binary directly, call binary_to_list if necessary
 read(This0, string) ->
   {This1, {ok, Sz}}  = read(This0, ui32),
-  read_data(This1, Sz).
+  read_data(This1, Sz, This0#t_compact.clone_binaries).
 
--spec read_data(#t_compact{}, non_neg_integer()) ->
+-spec read_data(#t_compact{}, non_neg_integer(), boolean()) ->
     {#t_compact{}, {ok, binary()} | {error, _Reason}}.
-read_data(This, 0) -> {This, {ok, <<>>}};
-read_data(This = #t_compact{transport = Trans}, Len) when is_integer(Len) andalso Len > 0 ->
-    {NewTransport, Result} = thrift_transport:read(Trans, Len),
+read_data(This, 0, _) -> {This, {ok, <<>>}};
+read_data(This, Len, Clone) when is_integer(Len) andalso Len > 0 ->
+    #t_compact{transport = Trans} = This,
+    {NewTransport, Result0} = thrift_transport:read(Trans, Len),
+    Result = case Result0 of
+                 {ok, Bin} when is_binary(Bin), Clone, Len > 64 ->
```

Comment:
The last condition should be removed. Cursory look at the BEAM code revealed that only brand new binaries can be allocated on heap, no copying is done for slices

Reply:
Are we missing a test case then? :)

Reply:
I'm not sure how one can reliably test this, as refc/heap binary dichotomy is hidden in the Erlang VM internals and there is no API to check which is which. The only (known to me) way to detect refc memory leak is through memory region monitoring, and it's beyond scope of unit/component tests. So regression testing is probably the best we can do.

Reply:
It sounded like you wanted to make one more change based on your original comment?

Reply:
Done